### PR TITLE
Fetching subdirectories of LinkResource

### PR DESCRIPTION
**Pull request by msojda**
Fixes puli/issues#129. 

As @tgalopin suggested I'm creating PR for fetching linked resources only.

### CODE REVIEW

File: src/InMemoryRepository.php [hunk 1]
```diff
@@ -63,11 +64,17 @@ public function get($path)
...
 
-        if (!isset($this->resources[$path])) {
-            throw ResourceNotFoundException::forPath($path);
+        if (isset($this->resources[$path])) {
+            return $this->resources[$path];
         }
 
-        return $this->resources[$path];
+        if ($link = $this->getLinkFromPath(Path::getDirectory($path))) {
+            $relativeTargetPath = ltrim($path, $link->getPath());
```
Comment (EmanueleMinotto) [inline src/InMemoryRepository.php]: The second parameter is a set of characters, this won't give the relative path.

File: src/InMemoryRepository.php [hunk 2]
```diff
@@ -281,6 +288,28 @@ private function getChildIterator(PuliResource $resource)
...
+            return;
+        }
+
+        if (!isset($this->resources[$path])) {
+            if ($link = $this->getLinkFromPath(Path::getDirectory($path))) {
+                return $link;
+            }
+        } elseif (isset($this->resources[$path]) && $this->resources[$path] instanceof LinkResource) {
+            return $this->resources[$path];
+        }
```
Comment (EmanueleMinotto) [inline src/InMemoryRepository.php]: You can reduce the complexity replacing this block with:
```php
if (!isset($this->resources[$path]) && $link = $this->getLinkFromPath(Path::getDirectory($path))) {
    return $link;
}

if (isset($this->resources[$path]) && $this->resources[$path] instanceof LinkResource) {
    return $this->resources[$path];
}
```

Comment (EmanueleMinotto) [inline src/InMemoryRepository.php]: You can reduce the complexity replacing with this block:
```php
if (isset($this->resources[$path]) && $this->resources[$path] instanceof LinkResource) {
    return $this->resources[$path];
}

if (!isset($this->resources[$path]) && $link = $this->getLinkFromPath(Path::getDirectory($path))) {
    return $link;
}
```

File: src/InMemoryRepository.php [hunk 3]
```diff
@@ -281,6 +288,28 @@ private function getChildIterator(PuliResource $resource)
     }
 
     /**
+     * Returns link from requested path if any found.
+     *
+     * @param string $path The path.
+     *
+     * @return LinkResource The link.
```
Comment (EmanueleMinotto) [inline src/InMemoryRepository.php]: If `'/' === $path` this method returns `null`, so I think should be added here.

File: src/InMemoryRepository.php (hunk 1 above)
Comment (tgalopin) [inline src/InMemoryRepository.php]: You can use `Path::makeRelative()`

**Comment by webmozart** [comment]
Thanks @msojda for working on this! :) Please don't forget to add tests that cover the different cases and edge cases of this functionality. Ideally, you should add these tests to the abstract test classes, so that the same tests are used for all repository implementations.

**Comment by webmozart** [comment]
What's the status here?

**Comment by tgalopin** [comment]
This aimed to fix https://github.com/puli/issues/issues/129 but became a dupplicate of https://github.com/puli/repository/pull/64 when we chose to not implement add and removal in links children. I think this can be closed, but it depends on the amount of work you did there @msojda : did you work a lot on fetching resources? It could be interesting to see how you did it and integrate some ideas in https://github.com/puli/repository/pull/64.
